### ISSUE TITLE
Support Custom TCP Stream Handling in SOCKS5 Proxy Mode

### ISSUE THREAD

**Issue by acyloxy** [mitmproxy/mitmproxy#7309]
#### Problem Description
Currently, in SOCKS5 proxy mode, mitmproxy lacks the capability for custom processing of TCP streams. This limitation presents challenges in handling cases where packet reassembly, encoding/decoding, encryption/decryption, and other stream manipulations are required. Such functionality would be especially beneficial for dealing with protocols that employ packet fragmentation or non-standard encoding mechanisms, where stream integrity and interpretability are crucial.

#### Proposal
Introduce a feature in mitmproxy’s SOCKS5 proxy mode that allows users to define custom handlers for TCP streams. These handlers could provide hooks or callbacks that facilitate processing for packet assembly/disassembly, encoding, and encryption tasks. By adding this feature, users could more effectively manage and manipulate TCP stream data, enhancing flexibility and control over complex data flows.

#### Alternatives
An alternative solution would be to implement this functionality at the application level, but this approach may lead to inefficiencies and increased complexity for users attempting to handle TCP stream processing outside of mitmproxy’s framework. Another approach could involve using external modules or plugins, though these options may not integrate seamlessly within mitmproxy’s existing structure.

#### Additional context
Adding this feature could improve mitmproxy’s versatility, particularly in environments requiring secure data handling, custom packet assembly, or handling non-standard protocols over TCP. Integrating custom TCP stream processing would align with mitmproxy’s goals of enhancing visibility and control over network traffic.

**Comment by mhils** [comment]
> Currently, in SOCKS5 proxy mode, mitmproxy lacks the capability for custom processing of TCP streams. 

That's not quite correct, `tcp_hosts` and friends should work in SOCKS mode as well.

https://docs.mitmproxy.org/stable/addons-examples/#tcp-simple has an example.

**Comment by acyloxy** [comment]
> > Currently, in SOCKS5 proxy mode, mitmproxy lacks the capability for custom processing of TCP streams.
> 
> That's not quite correct, `tcp_hosts` and friends should work in SOCKS mode as well.
> 
> https://docs.mitmproxy.org/stable/addons-examples/#tcp-simple has an example.

Yes, I have read the example you referenced above and learned that there are some basic functionalities to intercept and handle TCP streams such as data replacing, event hooking and so on.

Howerver, what I think here that can make the debugging capability over TCP streams more powerful is that `mitmproxy` can introduce packet reassembly towards the raw streams in order to identify the boundaries of each application packet.

**Comment by mhils** [comment]
I'm not sure I understand what you are asking for. TCP is fundamentally a stream-based protocol and not a message-based one. mitmproxy operates on sockets, i.e. above the transport layer. Individual packets are handled by the kernel.*


* Not true for WireGuard/local redirect mode, but that's a digression. There it's handled in Rust and mitmproxy still sees streams.

**Comment by lukant** [comment]
If I understand correctly, @acyloxy would like mitmproxy to split TCP streams to Application Layer messages. Right now `TCPMessages` are created for each read of a stream. Hence if peer will manage to send two messages, between two consecutive read, the `TCPMessage` will contain two packets. If peer will send messages in multiple packets, it is possible some `TCPMessage`s will contain only part of application message.

Addons that would like to process/amend tcp_messages will need to take this into account or they will fail from time to time.

There is a way to implement support for that without any change in `mitmproxy`. You may implement your own Layer that will hold data until whole packet is received and will split received data to multiple events in case more than one message was received at once.

The example below contains:
- an addon `MqttNextLayer` which make mitmproxy select `MqttLayer` if TCP connection is made to a default mqtt ports (1883 or 8883).
- the `MqttLayer` that extends `MessageReassemblyLayer` by defining its own `_get_packet_length` which is expected to return number of bytes that should be consider a TCPMessage or None if only part of message was received.
- `MessageReassemblyLayer` which extends standard mitmproxy's `TCPLayer`. It provides simplified interface for custom logic to detect messages boundries.

Beware that this will change the mitmproxy behavior slightly. Now, `mitmproxy` will send message to another peer only when whole message was received.


    from typing import Optional, Iterable
    import logging
    
    from mitmproxy.proxy import layer, events
    from mitmproxy.connection import Connection
    from mitmproxy.proxy.layers import ClientTLSLayer, TCPLayer
    from mitmproxy.proxy.layers.tcp import TcpMessageInjected
    from dataclasses import dataclass
    
    DEFAULT_MQTT_PORT = 1883
    DEFAULT_MQTT_TLS_PORT = 8883
    
    
    @dataclass
    class UncommittedData:
        received_data: bytes
        packets_to_be_injected: list[bytes]
    
    
    class MessageReassemblyLayer(TCPLayer):
        client = UncommittedData(b"", [])
        server = UncommittedData(b"", [])
    
        def _get_packet_length(self, buffer: bytes) -> Optional[int]:
            if len(buffer) > 0:
                return len(buffer)
    
            return None
    
        def _prepare_events(self, connection: Connection) -> Iterable[events.DataReceived]:
            uncommitted_data = self.client if connection == self.context.client else self.server
    
            while len(uncommitted_data.received_data) != 0:
                packet_size = self._get_packet_length(uncommitted_data.received_data)
    
                if not packet_size:
                    break
    
                yield events.DataReceived(connection, uncommitted_data.received_data[0:packet_size])
                uncommitted_data.received_data = uncommitted_data.received_data[packet_size:]
    
            for injected_event in uncommitted_data.packets_to_be_injected:
                yield events.DataReceived(connection, injected_event)
    
        def relay_messages(self, event: events.Event) -> layer.CommandGenerator[None]:
            connection : Optional[Connection] = None
    
            if isinstance(event, TcpMessageInjected):
                if event.message.from_client:
                    self.client.packets_to_be_injected.append(event.message.content)
                    connection = self.context.client
                else:
                    self.server.packets_to_be_injected.append(event.message.content)
                    connection = self.context.server
    
            if isinstance(event, events.DataReceived):
                connection = event.connection
                if connection == self.context.client:
                    self.client.received_data += event.data
                else:
                    self.server.received_data += event.data
    
            if connection is not None:
                for prepared_event in self._prepare_events(connection):
                    yield from super().relay_messages(prepared_event)
            else:
                yield from super().relay_messages(event)
    
    
    class MqttLayer(MessageReassemblyLayer):
        def _get_packet_length(self, buffer: bytes) -> Optional[int]:
            if len(buffer) < 2:
                return None
    
            decoder = iter(buffer[1:])
    
            multiplier = 1
            header_size = 1
            remaining_bytes_value = 0
    
            try:
                next_byte = decoder.__next__()
                header_size += 1
                remaining_bytes_value += (next_byte & 0x7F) * multiplier
                multiplier *= 0x80
                if next_byte & 0x80 == 0:
                    packet_length = remaining_bytes_value + header_size
                    logging.info(f"Packet size {packet_length} buffer: {buffer}")
                    if len(buffer) < packet_length:
                        return None
                    else:
                        return packet_length
            except StopIteration:
                return None
    
    class MqttNextLayer:
        def next_layer(self, nextlayer: layer.NextLayer):
            if nextlayer.layer:
                return  # do not override something another addon has set.
    
            if not nextlayer.context.server.address:
                return
    
            server_addr = nextlayer.context.server.address[1]
    
            def is_tls_already_established(context):
                return any(isinstance(l, ClientTLSLayer) for l in context.layers)
    
            if (server_addr == DEFAULT_MQTT_TLS_PORT and is_tls_already_established(nextlayer.context)) or \
                server_addr == DEFAULT_MQTT_PORT:
                nextlayer.layer = MqttLayer(nextlayer.context, ignore=False)
                return
    
    addons = [MqttNextLayer()]
